src/test/util/test-rig.ts

```diff
@@ -326,28 +326,35 @@ class ExecResult {
   /**
    * Waits for the given content to be logged to either stdout or stderr.
    *
-   * When it does, it consumes all the stdout and stderr that's been emitted
-   * so far and returns it.
+   * When it does, it consumes all stdout or stderr that's been emitted up to
+   * that match so far.
    */
-  async waitForLog(matcher: RegExp): Promise<{stdout: string; stderr: string}> {
+  waitForLog(matcher: RegExp): Promise<void> {
     const deferred = new Deferred<void>();
     this._logMatchers.push({re: matcher, deferred});
     // In case we've already received the log we're watching for
     this._checkMatchersAgainstLogs();
-    await deferred.promise;
-    const stdout = this._stdout;
-    const stderr = this._stderr;
-    this._stdout = '';
-    this._stderr = '';
-    return {stdout, stderr};
+    return deferred.promise;
   }
 
   private _checkMatchersAgainstLogs() {
-    for (const matcher of this._logMatchers) {
-      if (matcher.re.test(this._stdout) || matcher.re.test(this._stderr)) {
-        matcher.deferred.resolve();
+    let stdoutLastIndex = -1;
+    let stderrLastIndex = -1;
+    for (const {re, deferred} of this._logMatchers) {
+      if (re.test(this._stdout)) {
+        deferred.resolve();
+        stdoutLastIndex = Math.max(stdoutLastIndex, re.lastIndex);
```

Comment:
Hm yeah. Actually using `lastIndex` was completely wrong anyway, because it's only set when the `g` or `s` flags are on.

Switched to using `exec` (so that we always get one match even if the `g` flag was used), and checking the index + length, instead of lastIndex.

I also noticed we weren't removing matchers after they matched. We don't care about `lastIndex` now because it can only match once.